### PR TITLE
Issue #3306568 by vnech: Add ability for event managers to show/hide enrolees

### PR DESCRIPTION
**Pull request by nechai**
## Problem
In the event node type we have the enrollments page and block which are available for verified (authenticated) users.
Sometimes is useful to not show for user other participants who takes part in the event to avoid some side impact for her/his decision regarding participating.

## Solution
- Add a new field "Hide enrollments" allows to show/hide enrollers in the event.

## Issue tracker
- https://www.drupal.org/project/social/issues/3306568
- https://getopensocial.atlassian.net/browse/PROD-21834

## Theme issue tracker
N/A

## How to test
- [x] Login as SM
- [x] Create an event `Event #1`
- [x] Add a few participants (enrolles)
- [x] Login as LU
- [x] Visit event `Event #1` page
  - [x] EB: LU should _be able_ to see the "Enrollments" tab and visit this page as well
- [x] Login as SM
- [x] Edit the event `Event #1` and enable "Hide enrollments" option
- [x] Login as LU
- [x] Visit event `Event #1` page
  - [x] EB: LU should _not be able_ to see the "Enrollments" tab and visit this page as well

### Before merge
- [ ] Code/peer review is completed
- [ ] All commit messages are [clear and clean](https://open-social.slite.com/app/docs/DnmermZDIx_0OQ). If applicable a rebase was performed
- [ ] All automated tests are green
- [ ] Functional/manual tests of the acceptance criteria are approved
- [ ] All acceptance criteria were met
- [ ] New features or changes to existing features are covered by tests, either unit (preferably) or behat
- [ ] Update path is tested. New hook_updates should respect update order, right naming convention and consider hook_post_update code
- [ ] Module can be safely uninstalled. Update/implement hook_uninstall and make sure that removed configuration or dependencies are removed/uninstalled
- [ ] This pull request has all required labels (team/type/priority)
- [ ] This pull request has a milestone
- [ ] This pull request has an assignee (if applicable)
- [ ] Any front end changes are tested on all major browsers
- [ ] New UI elements, or changes on UI elements are approved by the design team
- [ ] New features, or feature changes are approved by the product owner

### After merge
- [ ] Code is tested on all branches that it has been cherry-picked
- [ ] Update hook number might need adjustment, make sure they have the correct order
- [ ] The Drupal.org ticket(s) are updated according to this pull request status

### A new option on event add/edit page:
Screenshot 2022-08-29 at 15 17 05

### The main event page when "Hide enrollments" is **disabled**:
Screenshot 2022-08-29 at 15 21 02

### The main event page when "Hide enrollments" is **enabled**:
Screenshot 2022-08-29 at 15 19 35

## Release notes
- *Add ability to hide enrollments on events*
- *Add node bundle class for event node type*

## Change Record (https://www.drupal.org/node/3307141)
- Introduce node bundle class for `event` node type ([see](https://www.drupal.org/node/3191609)) - `\Drupal\social_event\Entity\Node\Event`
- The sevice `social_event.enroll` is deprecates as main functionality was moved to event node bundle class `\Drupal\social_event\Entity\Node\Event`

## Translations
N/A

### CODE REVIEW

File: modules/social_features/social_event/config/install/field.field.node.event.field_hide_enrollments.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+dependencies:
+  config:
+    - field.storage.node.field_hide_enrollments
+    - node.type.event
+id: node.event.field_hide_enrollments
+field_name: field_hide_enrollments
+entity_type: node
+bundle: event
+label: 'Hide enrollments'
+description: 'Allows to restrict access to enrollments for users'
```
Comment (ribel) [inline modules/social_features/social_event/config/install/field.field.node.event.field_hide_enrollments.yml:12]: @nechai Did you confirm this copy with the Design team?

File: modules/social_features/social_event/social_event.install [hunk 2]
```diff
@@ -2163,3 +2163,35 @@ function social_event_update_11403(): string {
   // Output logged messages to related channel of update execution.
   return $updateHelper->logger()->output();
 }
+
+/**
+ * Add new "Hide enrollments" field to Event node type.
+ */
+function social_event_update_11404(): void {
```
Comment (ribel) [inline modules/social_features/social_event/social_event.install:2170]: @nechai Since this is a new feature, it should go to the next Major release - 11.5.0
That's why we need to start update hook numbers with `115xx`.

File: modules/social_features/social_event/social_event.module [hunk 3]
```diff
@@ -60,6 +61,13 @@ function social_event_theme(): array {
   ];
 }
 
+/**
+ * Implements hook_entity_bundle_info_alter().
+ */
+function social_event_entity_bundle_info_alter(array &$bundles): void {
```
Comment (ribel) [inline modules/social_features/social_event/social_event.module:67]: @nechai Please add explanation comment why this is needed.

File: modules/social_features/social_event/modules/social_event_an_enroll/social_event_an_enroll.module [hunk 4]
```diff
@@ -201,10 +202,10 @@ function social_event_an_enroll_menu_local_tasks_alter(&$data, $route_name) {
 function social_event_an_enroll_is_enabled(Node $node) {
   $config = \Drupal::config('social_event_an_enroll.settings');
   $is_global_enabled = $config->get('event_an_enroll');
-  $is_event = $node->getType() === 'event';
+  $is_event = $node instanceof Event;
   $is_public = $node->get('field_content_visibility')->getString() === 'public';
   $is_event_an_enroll = !empty($node->get('field_event_an_enroll')->value);
-  $is_event_enroll = \Drupal::service('social_event.enroll')->isEnabled($node);
+  $is_event_enroll = $is_event && $node->isEnrollmentEnabled();
```
Comment (ribel) [inline modules/social_features/social_event/modules/social_event_an_enroll/social_event_an_enroll.module:208]: @nechai Could you explain why this change was needed?

File: modules/social_features/social_event/social_event.services.yml [hunk 5]
```diff
@@ -19,3 +19,8 @@ services:
     class: Drupal\social_event\Service\SocialEventEnrollService
     arguments:
       - '@config.factory'
+    deprecated: The "%service_id%" service is deprecated in social:11.4.0 and is removed from social:12.0.0. See https://www.drupal.org/project/social/issues/3306568
```
Comment (ribel) [inline modules/social_features/social_event/social_event.services.yml:22]: @nechai I think we need to created separate Change Record for this deprecation.
Also it should go to 11.5.0 and 12.0.0 then.

File: modules/social_features/social_event/src/Service/SocialEventEnrollService.php [hunk 6]
```diff
@@ -34,28 +34,9 @@ public function __construct(ConfigFactoryInterface $config_factory) {
...
-    // This doesn't allow people to enroll themselves, but get invited.
-    if (
-      !$node->get('field_enroll_method')->isEmpty() &&
-      (int) $node->get('field_enroll_method')->getString() === EventEnrollmentInterface::ENROLL_METHOD_INVITE
-    ) {
-      $is_enabled = TRUE;
-    }
-
-    return $was_not_changed || $is_enabled;
+    @trigger_error(__METHOD__ . '() is deprecated in social:11.4.0 and is removed from social:12.0.0. Use bundled node object itself `$event->isEnrollmentEnabled()` instead. See https://www.drupal.org/project/social/issues/3306568', E_USER_DEPRECATED);
```
Comment (ribel) [inline modules/social_features/social_event/src/Service/SocialEventEnrollService.php:37]: @nechai Please change version from 11.4 to 11.5

File: modules/social_features/social_event/src/Service/SocialEventEnrollServiceInterface.php [hunk 7]
```diff
@@ -19,6 +19,10 @@ interface SocialEventEnrollServiceInterface {
    *
    * @return bool
    *   TRUE if enrollment is allowed.
+   *
+   * @deprecated in social:11.4.0 and is removed from social:12.0.0. Use
```
Comment (ribel) [inline modules/social_features/social_event/src/Service/SocialEventEnrollServiceInterface.php:23]: @nechai Please change version from 11.4 to 11.5

File: phpstan-baseline.neon [hunk 8]
```diff
@@ -17233,3 +17228,31 @@ parameters:
 			message: "#^Method Drupal\\\\social_flexible_group_book\\\\Plugin\\\\ActivityEntityCondition\\\\BookParentOnlyActivityEntityCondition\\:\\:isValidEntityCondition\\(\\) has parameter \\$entity with no type specified\\.$#"
 			count: 1
 			path: modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/src/Plugin/ActivityEntityCondition/BookParentOnlyActivityEntityCondition.php
+
+		-
+			message: "#^The \"social_event\\.enroll\" service is deprecated in social\\:11\\.4\\.0 and is removed from social\\:12\\.0\\.0\\. See https\\://github\\.com/goalgorilla/open_social/pull/3084$#"
```
Comment (ribel) [inline phpstan-baseline.neon:17233]: @nechai Please check PHPStan. Looks like we need to update more rules.

**Review by ribel — CHANGES_REQUESTED** [review]
@nechai I have tested functionality and it works good, but since this is a new feature it should to go to 11.5.0.
Also would be good to get more explanation about changes and deprecation in services. We need to create separate Change Record for it (make it unpublished for now).

File: modules/social_features/social_event/config/install/field.field.node.event.field_hide_enrollments.yml (hunk 1 above)
Comment (nechai) [reply modules/social_features/social_event/config/install/field.field.node.event.field_hide_enrollments.yml:12]: Confirmed with Design team

File: modules/social_features/social_event/social_event.module (hunk 3 above)
Comment (nechai) [reply modules/social_features/social_event/social_event.module:67]: Added

File: modules/social_features/social_event/modules/social_event_an_enroll/social_event_an_enroll.module (hunk 4 above)
Comment (nechai) [reply modules/social_features/social_event/modules/social_event_an_enroll/social_event_an_enroll.module:208]: This is refactoring change. As I added bundle class to event we can check if event have enrolments enabled inside an event object. Also, in my opinion  the method name `isEnrollmentEnabled()` has more sense rather then `isEnabled()` 🙂

File: modules/social_features/social_event/src/Service/SocialEventEnrollService.php (hunk 6 above)
Comment (nechai) [reply modules/social_features/social_event/src/Service/SocialEventEnrollService.php:37]: Added

File: modules/social_features/social_event/src/Service/SocialEventEnrollServiceInterface.php (hunk 7 above)
Comment (nechai) [reply modules/social_features/social_event/src/Service/SocialEventEnrollServiceInterface.php:23]: Added

File: phpstan-baseline.neon (hunk 8 above)
Comment (nechai) [reply phpstan-baseline.neon]: Added

**Review by ribel — APPROVED** [review]
Now all looks good to me.
@tbsiqueira or @navneet0693 could you please review it as well, since for this feature we also created a change record: https://www.drupal.org/node/3307141

File: modules/social_features/social_event/social_event.services.yml [hunk 9]
```diff
@@ -19,3 +19,8 @@ services:
     class: Drupal\social_event\Service\SocialEventEnrollService
     arguments:
       - '@config.factory'
+    deprecated: The "%service_id%" service is deprecated in social:11.5.0 and is removed from social:12.0.0. See https://www.drupal.org/project/social/issues/3306568
```
Comment (navneet0693) [inline modules/social_features/social_event/social_event.services.yml:22]: Nice!!! I didn't knew that we can do it here as well.

File: modules/social_features/social_event/social_event.install [hunk 10]
```diff
@@ -2163,3 +2163,35 @@ function social_event_update_11403(): string {
...
+/**
+ * Add new "Hide enrollments" field to Event node type.
+ */
+function social_event_update_11501(): void {
+  $config_path = \Drupal::service('extension.list.module')->getPath('social_event') . '/config/static';
+  $source = new FileStorage($config_path);
+
+  // Create a new field from the yaml configuration and save.
+  \Drupal::entityTypeManager()->getStorage('field_storage_config')
+    ->create((array) $source->read('field.storage.node.field_hide_enrollments_11501'))
```
Comment (navneet0693) [inline modules/social_features/social_event/social_event.install:2176]: Why we have chosen to type caste it as array ?

File: modules/social_features/social_event/social_event.module [hunk 11]
```diff
@@ -60,6 +61,20 @@ function social_event_theme(): array {
   ];
 }
 
+/**
+ * Implements hook_entity_bundle_info_alter().
+ *
+ * Set node bundle class for "Event" node type. This allows to encapsulate
```
Comment (navneet0693) [inline modules/social_features/social_event/social_event.module:67]: Please update the description to add what it encapsulates?

File: modules/social_features/social_event/social_event.module [hunk 12]
```diff
@@ -699,8 +714,14 @@ function social_event_block_access(Block $block, $operation, AccountInterface $a
   if ($operation === 'view' && ($block->getPluginId() === 'views_block:event_enrollments-event_enrollments')) {
     $nid = \Drupal::routeMatch()->getRawParameter('node');
     if (!empty($nid) && $node = Node::load($nid)) {
-      // Check if event enrollment were disabled explicitly.
-      if (!\Drupal::service('social_event.enroll')->isEnabled($node)) {
+      if (
+        // Show only for event node type.
```
Comment (navneet0693) [inline modules/social_features/social_event/social_event.module:718]: Nice commenting 👏🏾

File: modules/social_features/social_event/src/Access/EventEnrollmentsAccessCheck.php [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    if (
+      $account->hasPermission('bypass node access') ||
+      $account->hasPermission('administer event enrollment entities')
+    ) {
+      return AccessResult::allowed();
+    }
+
+    $node = $route_match->getParameter('node');
+    if (!$node instanceof NodeInterface) {
+      // On views url parameters node can be provided as raw parameter.
```
Comment (navneet0693) [inline modules/social_features/social_event/src/Access/EventEnrollmentsAccessCheck.php:34]: Would it possible to add actually relative paths so that developer can quickly recognize just by reading them ?

File: modules/social_features/social_event/src/Access/EventEnrollmentsAccessCheck.php [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+      $nid = $route_match->getRawParameter('node');
+      if (!empty($nid)) {
+        $node = Node::load($nid);
+      }
+    }
+
+    // Apply only for event node type.
+    if (!$node instanceof Event) {
+      // Returning `AccessResult::forbidden()` explicitly is better because for
```
Comment (navneet0693) [inline modules/social_features/social_event/src/Access/EventEnrollmentsAccessCheck.php:43]: Good documentations

File: modules/social_features/social_event/src/Entity/Node/Event.php [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+
+namespace Drupal\social_event\Entity\Node;
+
+use Drupal\social_event\EventEnrollmentInterface;
+use Drupal\social_node\Entity\Node as Node;
+
+/**
+ * Defines bundle class for event node type.
+ */
+final class Event extends Node implements EventInterface {
```
Comment (navneet0693) [inline modules/social_features/social_event/src/Entity/Node/Event.php:11]: So by choosing this to be a "final" class, are we right that we don't want any one to extend this?

File: modules/social_features/social_event/modules/social_event_invite/src/SocialEventInviteAccessHelper.php [hunk 16]
```diff
@@ -96,8 +96,18 @@ public function eventFeatureAccess() {
...
     // Get the group of this node.
-    $node = $this->routeMatch->getRawParameter('node');
-    $node = Node::load($node);
+    $node = $this->routeMatch->getParameter('node');
+    if (!$node instanceof NodeInterface) {
+      // On views url parameters node can be provided as raw parameter.
+      $nid = $this->routeMatch->getRawParameter('node');
+      // Node can't be loaded.
+      if (empty($nid)) {
+        return AccessResult::neutral();
```
Comment (navneet0693) [inline modules/social_features/social_event/modules/social_event_invite/src/SocialEventInviteAccessHelper.php:105]: Is this idea different from : https://github.com/goalgorilla/open_social/pull/3084/files#diff-19022dfe4a0998c73d0a6e27f8c20fd5c748ba1885cb3471b9e08e44e3aef3bcR42

File: modules/social_features/social_event/src/Entity/Node/Event.php [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+    $settings = \Drupal::config('social_event.settings');
+    if (
+      $settings->get('disable_event_enroll') ||
+      !$this->hasField('field_event_enroll') ||
+      (!$this->get('field_event_enroll')->isEmpty() && !$this->get('field_event_enroll')->getString())
+    ) {
+      return FALSE;
+    }
+
+    // When field was added to the event the value become `null` which
```
Comment (navneet0693) [inline modules/social_features/social_event/src/Entity/Node/Event.php:27]: Please update the description to say which "field" was added, also please improve the ending part. `if after saving the event the value is 'false'.`

File: modules/social_features/social_event/src/Entity/Node/Event.php [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+    // creates some inconsistent if after saving the event the value is `false`.
+    $was_not_changed = $this->get('field_event_enroll')->isEmpty();
+    $is_enabled = (bool) $this->get('field_event_enroll')->getString();
```
Comment (navneet0693) [inline modules/social_features/social_event/src/Entity/Node/Event.php:30]: Why are we typecasting?

**Review by navneet0693 — CHANGES_REQUESTED** [review]

File: modules/social_features/social_event/social_event.install (hunk 10 above)
Comment (nechai) [reply modules/social_features/social_event/social_event.install:2176]: Because otherwise we get `phpstan` checks failed.

File: modules/social_features/social_event/src/Entity/Node/Event.php (hunk 15 above)
Comment (nechai) [reply modules/social_features/social_event/src/Entity/Node/Event.php:11]: Yes, correct. Make class as final tells developers that `Event` behaviour should be only in `Event` itself - if you want to have something the same as `Event` just make double-think if you really need to have it. It save us  from having classes with similar behaviour which "extending" allows.

File: modules/social_features/social_event/src/Entity/Node/Event.php (hunk 17 above)
Comment (nechai) [reply modules/social_features/social_event/src/Entity/Node/Event.php:27]: added

File: modules/social_features/social_event/src/Entity/Node/Event.php (hunk 18 above)
Comment (nechai) [reply modules/social_features/social_event/src/Entity/Node/Event.php:30]: Because `getString()` returns `1` or `0` for boolean field, but we need to have `bool` type value to make possible having "strict type" conditions.

File: modules/social_features/social_event/src/Access/EventEnrollmentsAccessCheck.php (hunk 13 above)
Comment (nechai) [reply modules/social_features/social_event/src/Access/EventEnrollmentsAccessCheck.php:34]: I think it's bad idea to list here views paths (which are dynamic and I'm pretty sure it hard to do it here).

File: modules/social_features/social_event/modules/social_event_invite/src/SocialEventInviteAccessHelper.php (hunk 16 above)
Comment (navneet0693) [reply modules/social_features/social_event/modules/social_event_invite/src/SocialEventInviteAccessHelper.php:105]: @nechai The idea of returning AccessResults is a bit different from what I linked above. So, I wanted to know why there are differences in logic?

Comment (nechai) [reply modules/social_features/social_event/modules/social_event_invite/src/SocialEventInviteAccessHelper.php]: @navneet0693 Now I got what you mean - yes, definitely should be replaced with  `AccessResult::forbidden()` (but even current code will not break anything)

**Review by ribel — APPROVED** [review]

File: modules/social_features/social_event/src/Entity/Node/Event.php (hunk 15 above)
Comment (navneet0693) [reply modules/social_features/social_event/src/Entity/Node/Event.php:11]: @nechai One of the philosophy we want to follow in Open Social that we want our open source users to be able to extend the these classes. If they are extending, they have to be well aware of the development practices and what could go wrong. 

I do understand your approach and why you wanted to mark it as final, but we can leave upon at the discretion of community users/developers whether they want to extend it or not.

File: modules/social_features/social_event/modules/social_event_invite/src/SocialEventInviteAccessHelper.php (hunk 16 above)
Comment (navneet0693) [reply modules/social_features/social_event/modules/social_event_invite/src/SocialEventInviteAccessHelper.php]: @nechai You are right, this is not breaking anything :-D but I would to request to also place the same awesome idea logic here as well. The idea is really a good one.

File: modules/social_features/social_event/src/Entity/Node/Event.php (hunk 15 above)
Comment (nechai) [reply modules/social_features/social_event/src/Entity/Node/Event.php:11]: @navneet0693 removed `final` modifier 🙂

File: modules/social_features/social_event/modules/social_event_invite/src/SocialEventInviteAccessHelper.php (hunk 16 above)
Comment (nechai) [reply modules/social_features/social_event/modules/social_event_invite/src/SocialEventInviteAccessHelper.php]: @navneet0693 done

**Review by navneet0693 — APPROVED** [review]

**Comment by tbsiqueira** [comment]
@nechai , great work here, I loved the way you added the access checks and is using new Drupal 9.3 features, great documentation including release notes and change records.

Only missing feature are tests, but this is already aligned that we will do in a new branch

**Review by tbsiqueira — APPROVED** [review]

**Comment by ribel** [comment]
🍒 picked to 11.5.x
